jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/ExtractTarStep.java

```diff
@@ -139,21 +145,31 @@ public LocalImage call()
 
       // Process layer blobs
       // TODO: Optimize; compressing/calculating layer digests is slow
-      //       e.g. parallelize, faster compression method
+      //       e.g. faster compression method
       try (ProgressEventDispatcher progressEventDispatcher =
           progressEventDispatcherFactory.create(
               "processing base image layers", layerFiles.size())) {
         List<PreparedLayer> layers = new ArrayList<>(layerFiles.size());
         V22ManifestTemplate v22Manifest = new V22ManifestTemplate();
 
+        // Start compressing layers in parallel
+        List<Future<CachedLayer>> cachedLayers = new ArrayList<>();
         for (int index = 0; index < layerFiles.size(); index++) {
           Path layerFile = destination.resolve(layerFiles.get(index));
-          CachedLayer layer =
-              getCachedTarLayer(
-                  configurationTemplate.getLayerDiffId(index),
-                  layerFile,
-                  layersAreCompressed,
-                  progressEventDispatcher.newChildProducer());
+          DescriptorDigest diffId = configurationTemplate.getLayerDiffId(index);
+          ProgressEventDispatcher.Factory progressEventDispatcherFactory =
+              progressEventDispatcher.newChildProducer();
+
+          cachedLayers.add(
+              executorService.submit(
```

Comment:
Ideally, the code here won't deal with concurrency and `StepsRunner` should be able to maximize concurrency by individually compressing and caching each layer. For example, right now it is possible that some registry base image layers are being pulled in while some base layers can be pushed simultaneously. (I am sure constructing a base image manifest can be delayed, since I believe the reason we need them at an early stage is to be able to download the layers from a registry.) Also this would be nice for the consistent async framework design where each step doesn't get an `ExecutorService`.

But I think this is OK for short-term simple improvement for a low-hanging fruit. We should continue to track for the ideal async handling.

Reply:
Yeah, I agree this should be iterated on to be more consistent/less self-contained. But I figured it'd be nice to at least get the small performance gain in without adding too much complexity at the moment.

Reply:
Also if we change the build steps more it'd be nice to keep the second bullet in #1913 in mind.

>Instead of computing the layer digest and then pushing to a registry, push base image layers while compressing on-the-fly (#1906 (comment))

Reply:
Yeah, I'm already getting a headache over all these. Sounds like a fun challenge!